### DISCUSSION TITLE
Suggest new features!

### DISCUSSION THREAD

**Discussion by Lucas-C** [py-pdf/fpdf2#433] [Ideas]
`fpdf2` is a **living project**: it aims to grow with new features, improvements to existing functionalities, and always more helpful documentation.

To do so, we need to know what users of `fpdf2` would like to be able to do with this library.

This discussion purpose is exactly that: **please add comments with suggestions for features & general improvements!**

* would you like some [documentation](https://pyfpdf.github.io/fpdf2/) on how to use `fpdf2` in some way?
* would like to be able to produce some specific visuals?
* would you like support for a precise PDF feature?

<img src="https://camo.githubusercontent.com/6eb1daeb2251fb34a63c693634c34c0a1501b4ba504489ef33b0ceaeb41dc433/68747470733a2f2f6661767475746f722e636f6d2f7265736f75726365732f696d616765732f75706c6f6164732f313538383333363736353638325f626c6f675f31312d30315f2831295f2831292e6a7067" data-canonical-src="https://favtutor.com/resources/images/uploads/1588336765682_blog_11-01_(1)_(1).jpg" width="400" height="300" />

**Comment by vicmayrink** [comment]
FPDF is helping me to create high quality reports and documents. I really appreciate the community efforts to develop and maintain the project.

I would like to suggest a feature to include images in .pdf format. Most graphical libraries (including the famous matplotlib) allow us to save images as .pdf files. This a good choice to save the plot in vector format (rather than bitmaps). I know that FPDF2 supports SVG images, but I've got unexpected results when trying to embed an SVG image into the output PDF file.

**Comment by Lucas-C** [comment]
> I've got unexpected results when trying to embed an SVG image into the output PDF file

I'd be happy to help you properly embed SVG images in a PDF using `fpdf2`.
Could you please create an _issue_ and provide a bit more details on what was the problem exactly?

**Reply by vicmayrink** [reply]
Thanks for your reply :)
I've just created [this question](https://github.com/PyFPDF/fpdf2/issues/446)

**Comment by datatalking** [comment]
Does `fpdf2` have a `good first contribution` tag or whatever GitHub issues does ?

I'm looking to understand your framework as I used the old `PyFPDF` lib and had to patch the code, just found your repo and will have some time later this week to play with changes you've made.

Do we have a running list of functions or errors as a sort of diff between the last version `PyPDF2` and `fpdf2`?

When I import `fpdf2` why do some tutorials use a capitalized `FPDF2`, is this for a function difference? Perhaps the python importer care about capitalizations as it definitely cares wether its `pandas-dataframe` vs `pandas_dataframe` which I believe generates a `Import Error`

**Reply by Lucas-C** [reply]
> Does fpdf2 have a good first contribution tag or whatever GitHub issues does ?

We do have such issue tag indeed: [good first issue](https://github.com/PyFPDF/fpdf2/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22+sort%3Aupdated-desc)

> Do we have a running list of functions or errors as a sort of diff between the last version PyPDF2 and fpdf2?

So far I only know of only one functional difference with `PyFPDF`, detailed there: https://pyfpdf.github.io/fpdf2/Development.html#history

We aimed for full backward compatibility, however a few things have be deprecated (unused function arguments & other minor things), and will be removed one day.
You'll find instructions on how to enable deprecation warnings there: https://pyfpdf.github.io/fpdf2/index.html#displaying-deprecation-warnings

> why do some tutorials use a capitalized FPDF2?

I couldn't find a single use of this capitalized form in this repository, sorry.
Could you be more precise on what tutorials you are refering to?

`PyFPDF` is the original, unmaintained library: [GitHub repo](https://github.com/reingart/pyfpdf/) - [Pypi page](https://pypi.org/project/fpdf/)

`fpdf2` is this new, extended library: [GitHub repo](https://github.com/PyFPDF/fpdf2/) - [Pypi page](https://pypi.org/project/fpdf2/)

If you have any suggestion to make the difference between them clearer in our documentation, I'd be happy to hear about it 😊
(maybe open a new discussion about it)

**Reply by gmischler** [reply]
> So far I only know of only one functional difference with `PyFPDF`

There are several things that may technically still work, but are really deprecated in favour of cleaner solutions.
Not sure if all of the below were actually already present in the original PyFPDF, but they got changed at some time:

* `center` parameter for `cell()` is deprecated.
* `new_x` and `new_y` parameters to `cell()` and `multi_cell()` have replaced `ln`.
* `fill` parameters have been replaced by `style`.
* `dashed_line()` is deprecated in favour of `set_dash_pattern()`.
* parameter names have changed for `code39()` barcodes.
* `set_doc_option()` is deprecated.
* `dest` parameter for `output()` is deprecated.
* `type` parameter for `image()` is deprecated.
* `infile` parameter of `Template()` constructor is deprecated.
* Template elements have a transparent default background instead of white.
* `dest` parameter of `Template.render()` is deprecated.
* `rotate()` method is deprecated in favour of `rotation()` context manager.

The full set can of course be found in the changelog, but that may be tedious to scan through.
If you ask Python to show DeprecationWarnings, you'll also get a list of those that actually matter to your code.

But maybe a seperate list of API changes "since X" would really be helpful?
There should probably also be a plan about how long to carry around deprecated features, and when to actually remove them.

**Reply by Lucas-C** [reply]
> But maybe a seperate list of API changes "since X" would really be helpful?

Agreed!
Thank you for this list @gmischler.
We should mention those in the documentation, maybe in a dedicated section between "History" & "Usage" here:
https://pyfpdf.github.io/fpdf2/Development.html#history

Maybe like this:
```markdown
## History
## Compatibility with PyFPDF
### Major changes
### Deprecated functionalities
## Usage
```

What is your opinion on this proposal @gmischler?

I want to emphasize that the deprecated functionalities / arguments **still work perfectly fine**,
they just generate warnings suggesting to switch to alternative methods / parameters.

Hence 90% of the items in this list are deprecated functionalities that still work,
but the default background color for templates elements is something I did not have in mind that should probably be made explicit in the documentation.

> parameter names have changed for `code39()` barcodes

Which ones exactly? Is it just for templates?

**Reply by gmischler** [reply]
> Thank you for this list @gmischler.

No garantee that it's complete, though...

And of course, a lot of functionality has been added, but I guess in a migration context that only becomes relevant as a second step.

> What is your opinion on this proposal

Looks like a good place to put it.

> > parameter names have changed for `code39()` barcodes
> 
> Which ones exactly? Is it just for templates?

Ah yes, that was in the templates.

**Reply by Lucas-C** [reply]
I finally took the time to redact that PR 😅

Could you please review it @gmischler? -> https://github.com/py-pdf/fpdf2/pull/885

**Comment by n-gram-hub** [comment]
I would add an `add_page(bytes)` because other libraries (eg. PyPDF2) let you get (`get_contents().get_data()`) a byte string representation of a page. pdfrw lets you do a similar thing for a whole pdf in the PdfReader constructor, eg. reader = `PdfReader(fdata=bytes(fpdf.output()))`

**Reply by Lucas-C** [reply]
> I would add an `add_page(bytes)` because other libraries (eg. PyPDF2) let you get (`get_contents().get_data()`) a byte string representation of a page. pdfrw lets you do a similar thing for a whole pdf in the PdfReader constructor, eg. reader = `PdfReader(fdata=bytes(fpdf.output()))`

Hi @n-gram-hub! Thank you for submitting your suggestion 😊

If I understand correctly, you would like the ability to initiate a page with some byte content?
This is certainly doable, and I think that it could be added easily into the `contents=bytesarray()` created in `FPDF._beginpage()`:
https://github.com/PyFPDF/fpdf2/blob/2.7.4/fpdf/fpdf.py#L904

However, this will be a limited / low-level feature, as one will have to know that this `page.contents` does not include ALL content required to render a page. For example, if the page uses fonts & images, they won't be included, as those are inserted in dictionaries global to the whole PDF document.

What do you think about this @n-gram-hub?

**Reply by n-gram-hub** [reply]
Okay, I'll try to explain how this idea came about. A few days ago, using PyPDF2, I was looking for a way to edit a pdf: what I wanted to do was basically take an existing pdf, search for a string of text inside it, and create a new pdf that was identical but with this text edited. Truth be told, perhaps partly because I didn't have much time, I didn't succeed. However, I did notice this `get_contents().get_data()`, which returned something like:

`b'BT\n/TT1 1 Tf\n8 0 0 8 301.41 20.32 Tm\n/Cs6 cs 0.137 0.122 0.125 scn\n/GS1 gs\n0 Tc\n0 Tw\n(1)Tj\nET\nq\n1 i \n86.85 505.2 197.3 -150.2 re\nW n\nq\n197.3 0 0 150.2 86.85 355 cm\n/Im1 Do\nQ\nQ\n/Cs8 cs 0 scn\n297.5 30.7 14.2 -12.2 re\nf\n/Cs6 cs 0.863 0.867 0.871 scn\n527.15 804.45 34 -34.8 re\nf\nBT\n18 0 0 18 357.81 439.48 Tm\n0.137 0.122 0.125 scn\n0.0062 Tc\n(manual)Tj\n0 -1.2933 TD\n0.0057 Tc\n(\x92usage)Tj\n/TT2 1 Tf\n24 0 0 24 33.81 774.76 Tm\n-0.0077 Tc\n(Power)Tj\nET\nq\n57.9 0 0 59.3 507.25 404 cm\n/Im2 Do\nQ\nBT\n/TT1 1 Tf\n18 0 0 18 536.61 781.96 Tm\n-0.0553 Tc\n(EN)Tj\nET\n0.863 0.867 0.871 scn\n33.9 748.65 527.25 -34.8 re\nf\nBT\n17 0 0 17 43.17 726.16 Tm\n0.137 0.122 0.125 scn\n-0.0008 Tc\n-0.05 Tw\n(My text)Tj\nET\n/Cs6 CS 0.137 0.122 0.125 SCN\n0 J 0 j 0.96 w 10 M []0 d\n1 i \n33.81 61.36 m\n561.09 61.36 l\nS\nq\n72.4862 0 0 22.1 488.9 27.85 cm\n/Im3 Do\nQ\n'`

Here, one thing is clear: the text "My text" is in this string and any text is between (). So, I thought I could edit this string and create a page from this. I'm not a pdf expert and I don't know what kind of information is stored here. Of course, I understand what you are saying-I naively believed that all the information I needed to rebuild a page was in there.

**Reply by gmischler** [reply]
> A few days ago, using PyPDF2, I was looking for a way to edit a pdf: what I wanted to do was basically take an existing pdf, search for a string of text inside it, and create a new pdf that was identical but with this text edited.

This is a type of functionality (reading and editing PDFs) that fpdf2 explicitly does NOT want to support.
Compare: [What this library is not.](https://pyfpdf.github.io/fpdf2/FAQ.html#what-is-this-library-not)

Both [pdfrw](https://github.com/pmaupin/pdfrw) and [pypdf](https://github.com/py-pdf/pypdf) are likely better suited for such a task, if it is possible at all with reasonable effort. For background, the documentation of pypdf includes an explanation of [why text extraction is hard](https://pypdf.readthedocs.io/en/stable/user/extract-text.html#why-text-extraction-is-hard) on a relatively abstract level (it doesn't even touch on the more technical challenges). A simple text substitution could only possibly bring you success in the most trivial of use cases (replacing a single word, where the replacement is not longer than the original). Generally you might want to take a step back and consider if there is a way how you can solve the underlying task by a different approach, ideally applying at an earlier stage in the generation of your document.

**Reply by Lucas-C** [reply]
I agree with @gmischler overall.

A few months ago, I tried to write a script with zero dependency that performed text substitution on a PDF document:
https://github.com/Lucas-C/dotfiles_and_notes/blob/master/languages/python/pdf_edit_text.py

In the end, this script probably only works for the single PDF document I was working on,
and with serious limitations: it cannot insert letters that were not present initially in the document.

The functionality you were looking for @n-gram-hub, text edition in PDF documents, won't be provided by `fpdf2`.
Other libraries exist that can do that however! 😊

**Reply by n-gram-hub** [reply]
@gmischler First of all (sorry for suggesting something you don't want to support, I didn't know it) I must say, text extraction was a really easy task (maybe it's because I've been involved in text extracting activities before). 

Anyway, the problem was not directly mine: a friend, who works as a graphic designer, many years ago created a pdf with Adobe Illustrator and lost the source files: his main customer asked him to produce pdf translations of that file and he wanted to avoid creating a new manual from scratch. So, he asked for help and I devoted an hour and a half to him. Of course, he was inundated by three or four questions starting with "Why don't you...", but anything involving new work or a completely different approach to the problem was discarded. But in the end I gave up, as I said I'm not a pdf expert: I'm a curious guy, but the last thing I wanted to do was learning the inner workings of PyPDF2.   :-)

**Reply by Lucas-C** [reply]
Understood 😊

Even if we do not implement it, thanks for dropping by to suggest a new feature.
From a maintainer perspective, it's always interesting to know what users expect from the library.

**Reply by gmischler** [reply]
> I must say, text extraction was a really easy task (maybe it's because I've been involved in text extracting activities before).

If a document mainly consists of linear text, then extracting that can often be trivial.
But there's no garantee for that at all, because the PDF format doesn't have a concept of text flow.
There's a reason that specialized extractor programs exist that try to guess the format of tables in PDF files, with wildly varying success depending on what software created the file.

But we were talking about replacing text with a translation, which opens yet another can of worms...
If there is only a reasonably small amount of text involved, a graphical PDF editor might actually be best suited for the task. Acrobat Pro for example can edit text on screen, and your graphics designer friend might already have a licence of that. If there's a lot of files or a lot of text, the job might get a bit tedious, though...

**Reply by n-gram-hub** [reply]
@gmischler The problem with my friend is he doesn't want to do anything (manual editing, online APIs, nothing): the Python code he showed me was a fairly ridiculous code generated by a famous chatbot: he spent 10/15 minutes in the Socratic art of maieutics, believing something magic would have solved his issue   :-)   The same Adobe program he uses has this feature: https://www.christianda.com/blog-variable-data.html but he doesn't want to recreate the manual from scratch. Generating pdfs from data is not an impossible task, if you aren't that lazy.   :-)

I thank you for giving me a better understanding of the world around the pdf format (I knew it was intricate, but I didn't know how much)

**Comment by jhdedrick** [comment]
First, let me thank the authors and maintainers for a really exceptional piece of work.  This is quality software, and a real pleasure to use.

I'd like to request that you consider two additions to the .table() signature/behavior regarding the width of individual columns and the table overall.  The current behavior allows control of overall table width only if we're willing to live with equal-width columns, and allows specifying individual column widths only if we're willing to live with a fixed-width table.

Ideally, we would be able to take advantage of fpdf2's (excellent) column width apportionment logic most of the time, but be able to override it when needed.  And similarly, we'd like to be able to let the table use the entire width most of the time, but make it narrower or wider when needed.  Most importantly, we want both at the same time!

As background, I first need to point out an error in the documenation:

[Setting table & column widths¶](https://py-pdf.github.io/fpdf2/Tables.html#setting-table-column-widths)
> The col_widths optional parameter can be provided to configure this.
> If a single number is provided as col_widths, it is interpreted as a fixed column width in document units.
> If an array of numbers is provided as col_widths, the values are considered to be fractions of the full effective page width, meaning that col_widths=(1, 1, 2) is strictly equivalent to col_widths=(25, 25, 50).

(In fact, when a single number is provided, it actually becomes the total width of the table, not just of a column.  In this case, all column widths are the same, and are sized to fit within the width.)

The current signature has a single parameter, "col_widths".  If col_widths is a single value, it uses that value in document units for _every_ column.  If it's an iterable, instead of the expected behavior (using the values in document units to control the widths of columns individually), it treats them as relative sizes, but _resets the overall width to the effective document width_.  So, you can either control overall width or relative column size, _but not both_.

Detailed proposal (but we'd be happy with any implementation that gives the same degree of control):
1. Add a "table_width" parameter that defaults to the effective document width.  This takes a single float, and all tables are then exactly this width.  Deprecate the single-value input option for "column_widths", since it doesn't really do what it says it does.
2. Add a "col_abs_widths" parameter.  This is an iterable of float | None, which must be of length <= the number of columns.  If any value is specified, that value is the absolute width of the corresponding column in document units.  If a value is None or not specified, the existing fpdf2 column-width logic controls the width of that column (while staying within "table_width" of course.  Note that there then may be a mix of absolutely specified columns and columns controlled by fpdf2.
3. "col_widths" becomes effectively "col_rel_widths" - it takes an iterable of float | int | None, and acts like col_abs_widths, except with relative, not absolute values.

**Reply by andersonhc** [reply]
We already have a `width` parameter, which already behaves like the proposed `table_width`.
If you pass `width`, any `col_widths` sequence is treated as relative weights *within that width*.
If you don’t pass `width`, it defaults to the effective page width.

We don’t currently support absolute per‑column widths directly (or mixing fixed + auto),
but you can get exact absolute widths by converting them to ratios yourself
(e.g., set `width=sum(col_widths)` and pass the list as weights).

If it still doesn’t work for you, please open an issue with a minimal example so we can reproduce it.
